gk/main.c

```diff
@@ -16,22 +16,310 @@
+	if (priority <= fe->u.request.last_priority &&
+			fe->u.request.allowance > 0) {
+		fe->u.request.allowance--;
+		priority = fe->u.request.last_priority;
+	} else {
+		fe->u.request.last_priority = priority;
+		fe->u.request.allowance = START_ALLOWANCE - 1;
+	}
+
+	if (fe->u.request.allowance <= 0) {
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"gk: the allowance for the request is 0, need to be dropped!\n");
+		return drop_packet(packet->pkt);
+	}
+
+	/*
+	 * Adjust @priority for the DSCP field.
+	 * DSCP 0 for legacy packets; 1 for granted packets; 
+	 * 2 for capability renew; 3-63 for requests.
+	 */
+	priority += 3;
+	if (unlikely(priority > 63))
+		priority = 63;
+
+	/* The assigned priority is @priority. */
+	
+	/* TODO Encapsulate the packet as a request. */
+
+	/* TODO Put this encapsulated packet in the request queue. */
+
+	return 0;
+}
+
+static uint64_t
+time_from_seconds(uint64_t time)
+{
+	return (cycles_per_sec * time);
+}
+
+static int
+gk_process_granted(struct flow_entry *fe, struct ipacket *packet)
+{
+	bool renew_cap;
+	uint64_t now = rte_rdtsc();
+	struct rte_mbuf *pkt = packet->pkt;
+
+	/* 
+	 * When a flow entry is at granted state, the GK block processes 
+	 * that entry as follows:
+	 */
+	if (now >= fe->u.granted.cap_expire_at) {
+		initialize_flow_entry(fe, &packet->flow);
+		return gk_process_request(fe, packet);
+	}
+
+	if (now >= fe->u.granted.budget_renew_at) {
+		fe->u.granted.budget_renew_at = now + time_from_seconds(1);
+		fe->u.granted.budget_byte = fe->u.granted.tx_rate_kb_cycle * 1024;
+	}
+
+	if (pkt->data_len > fe->u.granted.budget_byte)
+		return drop_packet(pkt);
+
+	fe->u.granted.budget_byte -= pkt->data_len;
+	renew_cap = now >= fe->u.granted.send_next_renewal_at;
+	if (renew_cap)
+		fe->u.granted.send_next_renewal_at = now +
+			fe->u.granted.renewal_step_cycle;
+
+	/*
+	 * TODO Encapsulate packet as a granted packet,
+	 * mark it as a capability renewal request if @renew_cap is true,
+	 * enter destination according to @fe->u.granted.grantor_id,
+	 * and put the encapsulated packet in the granted queue.
+	 */
+
+	return 0;
+}
+
+static int
+gk_process_declined(struct flow_entry *fe, struct ipacket *packet)
+{
+	uint64_t now = rte_rdtsc();
+
+	/*
+	 * When a flow entry is at declined state, the GK block processes
+	 * that entry as follows:
+	 */
+	if (unlikely(now >= fe->u.declined.expire_at)) {
+		initialize_flow_entry(fe, &packet->flow);
+		return gk_process_request(fe, packet);
+	}
+
+	return drop_packet(packet->pkt);
+}
+
+static int
+setup_gk_instance(unsigned int block_idx, struct gk_config *gk_conf)
+{
+	int  ret;
+	char ht_name[64];
+
+	struct gk_instance *instance = &gk_conf->instances[block_idx];
+	struct rte_hash_parameters ip_flow_hash_params = {
+		.name = NULL,
+		.entries = gk_conf->flow_ht_size,
+		.key_len = sizeof(struct ip_flow),
+		.hash_func = DEFAULT_HASH_FUNC,
+		.hash_func_init_val = 0,
+	};
+
+	ret = snprintf(ht_name, sizeof(ht_name), "ip_flow_hash_%d", block_idx);
```

Comment:
If `block_idx` is an unsigned int quantity, the format specifier should be `%u`. If it's a `uint32_t`, it should be ` %"PRIu32"`.